tests/tests_pytorch/strategies/test_sharded_strategy.py

```diff
@@ -18,6 +20,37 @@
     from fairscale.optim import OSS
 
 
+class ModelWithAdamOptimizer(BoringModel):
+    def configure_optimizers(self):
+        optimizer = torch.optim.Adam(self.layer.parameters(), lr=0.1)
+        return optimizer
+
+
+class CheckModelRestore(ModelWithAdamOptimizer):
+    def __init__(self, old_model_state_dict, old_optimizer_states):
+        super().__init__()
+        self.old_model_state_dict = old_model_state_dict
+        self.old_optimizer_states = old_optimizer_states
+
+    def on_train_start(self):
```

Comment:
We discussed this briefly in the past offline in general. 

We will have to provide a standardized way to do this in Lightning. Not necessarily automatically on resuming, but at least by providing utilities in some form to do this, not just for Zero1, but for all types of sharded state dicts and checkpoints. 

I will look into making a GitHub issue.

Reply:
Apart from deepspeed, we don't support saving a shard of a checkpoint on each rank in any of the sharded strategyes. Issues like this one for example exist #11361. Once we support that, we will need to provide ways of resharding these checkpoints in an efficient way. 

Imagine you have a shard of N checkpoint files because you trained with N processes. Now you want to load these checkpoints into a trainer that runs with M != N processes.

The inefficient way we do it currently is to call consolidate_state_dict, but this will have to hold all shard in memory which is often not possible with the biggest models today.

Reply:
yep! agreed!

but is that blocking this PR? I understand we need to consolidate it first, so once we have a better way, we could update this test.

if you think we should close it, it's fine too :)

Reply:
No it's not blocking. It's just making aware that we are looking into options to generalize this and make it memory efficient.